napari/layers/image/_slice.py

```diff
@@ -0,0 +1,162 @@
+    slice_indices : Tuple[Union[int, slice], ...]
+        The slice indices in the layer's data space.
+    lazy : bool
+        If True, do not materialize the data with `np.asarray` during execution.
+        Otherwise, False. This should be True for the experimental async code
+        (as the load occurs on a separate thread) but False for the new async
+        where `execute` is expected to be run on a separate thread.
+    others
+        See the corresponding attributes in `Layer` and `Image`.
+    """
+
+    dims: _SliceInput
+    data: Any = field(repr=False)
+    slice_indices: Tuple[Union[int, slice], ...]
+    multiscale: bool = field(repr=False)
+    corner_pixels: np.ndarray
+    rgb: bool = field(repr=False)
+    data_level: int = field(repr=False)
+    thumbnail_level: int = field(repr=False)
+    level_shapes: np.ndarray = field(repr=False)
+    downsample_factors: np.ndarray = field(repr=False)
+    lazy: bool = field(default=False, repr=False)
+
+    def execute(self) -> _ImageSliceResponse:
+        return (
+            self._execute_multi_scale()
+            if self.multiscale
+            else self._execute_single_scale()
+        )
+
+    def _execute_single_scale(self) -> _ImageSliceResponse:
+        image = self.data[self.slice_indices]
+        if not self.lazy:
+            image = np.asarray(image)
+        # `Layer.multiscale` is mutable so we need to pass back the identity
+        # transform to ensure `tile2data` is properly set on the layer.
+        ndim = self.dims.ndim
+        tile_to_data = Affine(
+            name='tile2data', linear_matrix=np.eye(ndim), ndim=ndim
+        )
+        return _ImageSliceResponse(
+            data=image,
+            thumbnail=None,
+            tile_to_data=tile_to_data,
+        )
+
+    def _execute_multi_scale(self) -> _ImageSliceResponse:
+        if self.dims.ndisplay == 3:
+            warnings.warn(
+                trans._(
+                    'Multiscale rendering is only supported in 2D. In 3D, only the lowest resolution scale is displayed',
+                    deferred=True,
+                ),
+                category=UserWarning,
+            )
+            level = len(self.data) - 1
+        else:
+            level = self.data_level
+
+        indices = self._slice_indices_at_level(level)
+
+        # Calculate the tile-to-data transform.
+        scale = np.ones(self.dims.ndim)
+        for d in self.dims.displayed:
+            scale[d] = self.downsample_factors[level][d]
+
+        translate = np.zeros(self.dims.ndim)
+        if self.dims.ndisplay == 2:
+            for d in self.dims.displayed:
+                indices[d] = slice(
+                    self.corner_pixels[0, d],
+                    self.corner_pixels[1, d],
+                    1,
+                )
+            translate = self.corner_pixels[0] * scale
+
+        # This only needs to be a ScaleTranslate but different types
```

Comment:
Is it useful for us to maintain our own transform implementation?

Reply:
Do you mean our own transform implementation for slicing? Maybe, but I think it is a requirement that whatever we return should be easily convertible into a transform that napari can use.